### PR TITLE
[Cache] Support compilable cache reuse with smaller batch sizes

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

⚠️ this PR needs to be rebased, don't review/merge

Supercedes #37389 
Partially solves #35444

This PR makes our `max_cache_size` argument in compilable caches finally true: we can now use a cache object with a batch size smaller than the one defined in the cache. Compile once and run with multiple input shapes -- particularly useful for export, as mentioned in #35444.

Adds other minor related fixes (see PR comments).

__________________

We can see in the following test script that this does not degrade compiled performance:

```py
from transformers import AutoModelForCausalLM, AutoTokenizer, StaticCache
import torch
import time

tokenizer = AutoTokenizer.from_pretrained("meta-llama/Llama-3.2-1B")
model = AutoModelForCausalLM.from_pretrained("meta-llama/Llama-3.2-1B", device_map="auto", torch_dtype=torch.float16)

input_ids = tokenizer(["The quick brown"], return_tensors="pt").input_ids.to(model.device)
cache_position = torch.arange(input_ids.shape[1]).to(model.device)

with torch.no_grad():
    #------------------------------------------------------------------------------------------------
    # OLD, cache batch size = input batch size
    # Measured on an RTX 4090: `main` = 0.223ms; this PR = 0.223ms
    cache = StaticCache(
        config=model.config,
        max_batch_size=1,
        max_cache_len=100,
        device=model.device,
        dtype=model.dtype
    )
    model.forward = torch.compile(model.forward, fullgraph=True, mode="reduce-overhead")

    # warmup
    for _ in range(3):
        outputs = model(input_ids, cache_position=cache_position, past_key_values=cache)

    # measure
    start = time.time()
    for _ in range(100):
        outputs = model(input_ids, cache_position=cache_position, past_key_values=cache)
    end = time.time()
    print(f"[Old] Average time taken: {((end - start) / 100) * 1000} ms")

    #------------------------------------------------------------------------------------------------
    # clear torch compile cache
    torch._dynamo.reset()

    #------------------------------------------------------------------------------------------------
    # NEW, cache batch size > input batch size
    # Measured on an RTX 4090: `main` = Doesn't work; this PR = 0.224ms
    cache = StaticCache(
        config=model.config,
        max_batch_size=16,  # 16 >> 1
        max_cache_len=100,
        device=model.device,
        dtype=model.dtype
    )
    model.forward = torch.compile(model.forward, fullgraph=True, mode="reduce-overhead")

    # warmup
    for _ in range(3):
        outputs = model(input_ids, cache_position=cache_position, past_key_values=cache)

    # measure
    start = time.time()
    for _ in range(100):
        outputs = model(input_ids, cache_position=cache_position, past_key_values=cache)
    end = time.time()
    print(f"[New] Average time taken: {((end - start) / 100) * 1000} ms")

```

### CODE REVIEW

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -2014,10 +2015,12 @@ def reset(self):
 
 class MambaCache:
     """
-    Cache for mamba model which does not have attention mechanism and key value states.
+    Cache for mamba model which does not have attention mechanism and key value states.  At initialization, the cache
+    is preallocated to its maximum possible shape. Contrarily to other caches, `max_batch_size` must match the
+    batch size used at inference time.
```
Comment (gante) [inline src/transformers/cache_utils.py:2020]: Note: adding the feature to `MambaCache` would imply adding model-level changes. `mamba` doesn't have a lot of usage to justify the work

File: tests/utils/test_cache_utils.py [hunk 2]
```diff
@@ -285,8 +286,8 @@ def test_static_cache_exportability(self):
 
 
 @require_torch_accelerator
-@slow
```
Comment (gante) [inline tests/utils/test_cache_utils.py:288]: Not all tests here are `@slow`. Tests that take >1s kept the decorator

File: tests/utils/test_cache_utils.py [hunk 3]
```diff
@@ -479,44 +485,7 @@ def test_static_cache_greedy_decoding_pad_left(self, attn_implementation, cache_
...
             self.assertListEqual(decoded, EXPECTED_GENERATION)
 
-    @require_torch_gpu
-    @parameterized.expand(
-        [
-            ("eager", "static"),
-            ("sdpa", "static"),
-        ]
-    )
-    def test_static_cache_greedy_decoding_pad_right(self, attn_implementation, cache_implementation):
```
Comment (gante) [inline tests/utils/test_cache_utils.py:489]: generation is not meant to work well with right-padding, no need to spend resources testing it

**Review by zucchini-nlp — APPROVED** [review]
Great, thanks for adding support on this and happy to see it can be done with minimal changes!

**Comment by gante** [comment]
(PR on hold: some slow cache tests are failing, fixing them first before re-requesting a review)

**Comment by gante** [comment]
(caches have been refactored, better start from scratch)
